File: scripts/fetch_gpt_meta.py
import json
import sys
import os
from pathlib import Path
from openai import OpenAI

# -----------------------------
# Step 1: Get JSON file path
# -----------------------------
def get_json_file_path():
    """
    Retrieves the JSON file path from the command line arguments.
    Exits the script if no argument is provided or the file does not exist.
    Returns:
        Path object pointing to the JSON file.
    """
    if len(sys.argv) < 2:
        print("Usage: python fetch_gpt_meta.py <path_to_json_file>")
        sys.exit(1)

    json_file_path = Path(sys.argv[1])
    if not json_file_path.is_file():
        print(f"File not found: {json_file_path}")
        sys.exit(1)

    return json_file_path

json_file_path = get_json_file_path()

# -----------------------------
# Step 2: Load config and OpenAI API key
# -----------------------------
APP_DATA_DIR = Path(os.getenv("APPDATA", ".")) / "metadata-filler-v3"
CONFIG_PATH = APP_DATA_DIR / "config.json"

if not CONFIG_PATH.is_file():
    print(f"Config file not found: {CONFIG_PATH}")
    sys.exit(1)

with open(CONFIG_PATH, "r", encoding="utf-8") as f:
    config = json.load(f)

api_key = config.get("openai_credentials", {}).get("api_key")
if not api_key:
    print("API key not found in config file.")
    sys.exit(1)

client = OpenAI(api_key=api_key)

# -----------------------------
# Step 3: Load input JSON
# -----------------------------
with open(json_file_path, "r", encoding="utf-8") as f:
    data = json.load(f)

# -----------------------------
# Step 3b: Merge tracklist into Tracks if needed
# -----------------------------
if "tracklist" in data:
    if "Tracks" not in data:
        data["Tracks"] = data["tracklist"]
    else:
        # Merge without duplicates
        existing_urls = {t.get("url") for t in data["Tracks"]}
        for track in data["tracklist"]:
            if track.get("url") not in existing_urls:
                data["Tracks"].append(track)
    # Remove the old tracklist key
    del data["tracklist"]

# -----------------------------
# Step 4: Query GPT for album metadata
# -----------------------------
prompt = f"""
You are a strict music metadata assistant.

INSTRUCTIONS:
- Fill in all missing metadata for the album.
- Return ONLY a single valid JSON object with these exact keys:
contributing_artist, genre, rating, comments, publisher, encoded_by,
parental_rating_reason, composers, conductors, group_description, mood, part_of_set, protected.
- Do not include any explanation, notes, or extra text outside the JSON.
- Leave empty strings "" or zeros 0 for unknown values.
- Lists such as contributing_artist, composers, or conductors must be returned as JSON arrays.
- Keep the JSON valid and parsable.

EXISTING DATA:
{json.dumps(data)}
"""

try:
    response = client.chat.completions.create(
        model="gpt-5-mini",
        messages=[
            {"role": "system", "content": "You are a helpful metadata assistant."},
            {"role": "user", "content": prompt}
        ]
    )
    raw_content = response.choices[0].message.content.strip()
    start = raw_content.find("{")
    end = raw_content.rfind("}") + 1
    if start == -1 or end == -1:
        print("Warning: No JSON found for album")
        album_metadata = {}
    else:
        album_metadata = json.loads(raw_content[start:end])
except Exception as e:
    print(f"Error fetching album metadata: {e}")
    album_metadata = {}

# Merge album metadata safely
data.update({
    "contributing_artist": album_metadata.get("contributing_artist", []),
    "genre": album_metadata.get("genre", data.get("Genre", "")),
    "rating": album_metadata.get("rating", 0),
    "comments": album_metadata.get("comments", ""),
    "publisher": album_metadata.get("publisher", ""),
    "encoded_by": album_metadata.get("encoded_by", ""),
    "parental_rating_reason": album_metadata.get("parental_rating_reason", ""),
    "composers": album_metadata.get("composers", []),
    "conductors": album_metadata.get("conductors", []),
    "group_description": album_metadata.get("group_description", ""),
    "mood": album_metadata.get("mood", ""),
    "part_of_set": album_metadata.get("part_of_set", 0),
    "protected": album_metadata.get("protected", False),
})

# -----------------------------
# Step 5: Save updated JSON
# -----------------------------
with open(json_file_path, "w", encoding="utf-8") as f:
    json.dump(data, f, ensure_ascii=False, indent=4)

print(f"Album metadata updated and saved to {json_file_path}")

File: scripts/fetch_gpt_meta.js
/**
 * Fetch GPT metadata for tracks (Node.js replacement for Python OpenAI API)
 * Uses: openai npm package
 */

import fs from 'fs';
import path from 'path';
import { OpenAI } from 'openai';

/**
 * Extract metadata using OpenAI GPT
 */
async function fetchMetadataWithGPT(manifestPath, apiKey) {
    try {
        if (!fs.existsSync(manifestPath)) {
            console.error(`[GPT] Manifest not found: ${manifestPath}`);
            return;
        }

        if (!apiKey) {
            console.warn('[GPT] OpenAI API key not configured. Skipping GPT metadata extraction.');
            return;
        }

        const manifestData = JSON.parse(fs.readFileSync(manifestPath, 'utf8'));
        const tracks = manifestData.Tracks || [];

        console.log(`[GPT] Processing ${tracks.length} tracks for metadata...`);

        // Initialize OpenAI client
        const client = new OpenAI({ apiKey });

        // Prepare album-level metadata
        let albumMetadata = {};

        // Query GPT once for the album as a whole
        try {
            const albumPrompt = `Extract comprehensive music metadata for the following album:
Album Title: ${manifestData.Album_Title || 'Unknown'}
Artist: ${manifestData.Primary_Artist || 'Unknown'}
Release Date: ${manifestData.Release_Date || 'Unknown'}
Track Count: ${manifestData.Track_Count || 0}

Provide ONLY a valid JSON object (no markdown, no extra text):
{
  "genre": "primary genre",
  "mood": "overall mood/vibe",
  "publisher": "publisher/label name",
  "composers": ["composer1", "composer2"],
  "conductors": ["conductor1"],
  "group_description": "band/group description",
  "rating": 0,
  "comments": "album notes",
  "contributing_artist": ["artist1", "artist2"],
  "parental_rating_reason": "reason if applicable"
}`;

            console.log(`[GPT] Fetching album-level metadata...`);
            const albumMessage = await client.chat.completions.create({
                model: 'gpt-3.5-turbo',
                max_tokens: 1024,
                messages: [
                    {
                        role: 'user',
                        content: albumPrompt
                    }
                ]
            });

            const albumResponse = albumMessage.choices[0].message;
            if (albumResponse.content) {
                try {
                    albumMetadata = JSON.parse(albumResponse.content);
                    console.log(`[GPT] ✓ Album metadata: ${albumMetadata.genre} - ${albumMetadata.mood}`);
                } catch (e) {
                    console.warn(`[GPT] Could not parse album metadata response: ${e.message}`);
                }
            }
        } catch (err) {
            console.warn(`[GPT] Error fetching album metadata: ${err.message}`);
        }

        // Process individual tracks
        for (const track of tracks) {
            try {
                const trackPrompt = `Extract music metadata for this specific track:
Album: ${manifestData.Album_Title || 'Unknown'}
Artist: ${manifestData.Primary_Artist || 'Unknown'}
Track: ${track.title || 'Unknown'}
Duration: ${track.duration || 'Unknown'}

Provide ONLY a valid JSON object (no markdown, no extra text) with track-specific fields only:
{
  "isrc": "ISRC code if available"
}`;

                const message = await client.chat.completions.create({
                    model: 'gpt-3.5-turbo',
                    max_tokens: 256,
                    messages: [
                        {
                            role: 'user',
                            content: trackPrompt
                        }
                    ]
                });

                const response = message.choices[0].message;
                if (response.content) {
                    try {
                        const metadata = JSON.parse(response.content);
                        console.log(`[GPT] ✓ ${track.title}`);
                        
                        // Update track with ONLY track-specific metadata
                        track.isrc = metadata.isrc || '';
                    } catch (e) {
                        console.warn(`[GPT] Could not parse track response for ${track.title}: ${e.message}`);
                    }
                }
            } catch (err) {
                console.warn(`[GPT] Error processing track "${track.title}": ${err.message}`);
                // Continue to next track on error
            }
        }

        // Merge album-level metadata into manifest root using template field names
        if (Object.keys(albumMetadata).length > 0) {
            manifestData.genre = albumMetadata.genre || '';
            manifestData.mood = albumMetadata.mood || '';
            manifestData.contributing_artist = albumMetadata.contributing_artist || [];
            manifestData.rating = albumMetadata.rating || 0;
            manifestData.comments = albumMetadata.comments || '';
            manifestData.publisher = albumMetadata.publisher || '';
            manifestData.composers = albumMetadata.composers || [];
            manifestData.conductors = albumMetadata.conductors || [];
            manifestData.group_description = albumMetadata.group_description || '';
            manifestData.parental_rating_reason = albumMetadata.parental_rating_reason || '';
        }

        // Save updated manifest
        fs.writeFileSync(manifestPath, JSON.stringify(manifestData, null, 2));
        console.log(`[GPT] ✅ Metadata extraction complete`);

    } catch (err) {
        console.error(`[GPT] Fatal error: ${err.message}`);
    }
}

// Run if called directly
if (process.argv[1] === new URL(import.meta.url).pathname) {
    const manifestPath = process.argv[2];
    const apiKey = process.env.OPENAI_API_KEY;

    if (!manifestPath) {
        console.error('Usage: node fetch_gpt_meta.js <manifest_path>');
        console.error('Set OPENAI_API_KEY environment variable for GPT functionality');
        process.exit(1);
    }

    fetchMetadataWithGPT(manifestPath, apiKey);
}

export { fetchMetadataWithGPT };
